gen_version.py

```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env python
+#
+# Copyright 2015 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generate version string for NotoColorEmoji.
+
+This parses the color emoji template file and updates the lines
+containing version string info, writing a new file.
+
+The nameID 5 field in the emoji font should reflect the commit/date
+of the repo it was built from.  This will build a string of the following
+format:
+  Version 1.39;GOOG;noto-emoji:20170220:a8a215d2e889'
+
+This is intended to indicate that it was built by Google from noto-emoji
+at commit a8a215d2e889 and date 20170220 (since dates are a bit easier
+to locate in time than commit hashes).
+
+For building with external data we don't include the commit id as we
+might be using different resoruces.  Instead the version string is:
+  Version 1.39;GOOG;noto-emoji:20170518;BETA <msg>
+
+Here the date is the current date, and the message after 'BETA ' is
+provided using the '-b' flag.  There's no commit hash.  This also
+bypasses some checks about the state of the repo.
+
+The relase number should have 2 or 3 minor digits.  Right now we've been
+using 2 but at the next major relase we probably want to use 3.  This
+supports both.  It will bump the version number if none is provided,
+maintaining the minor digit length.
+"""
+
+import argparse
+import datetime
+import re
+
+from nototools import tool_utils
+
+# These are not very lenient, we expect to be applied to the noto color
+# emoji template ttx file which matches these.  Why then require the
+# input argument, you ask?  Um... testing?
+_nameid_re = re.compile(r'\s*<namerecord nameID="5"')
+_version_re = re.compile(r'\s*Version\s(\d+.\d{2,3})')
+_headrev_re = re.compile(r'\s*<fontRevision value="(\d+.\d{2,3})"/>')
+
+def _get_existing_version(lines):
+  """Scan lines for all existing version numbers, and ensure they match.
+  Return the matched version number string."""
+
+  version = None
+  def check_version(new_version):
+    if version is not None and new_version != version:
+      raise Exception(
+          'version %s and namerecord version %s do not match' % (
+              version, new_version))
+    return new_version
+
+  saw_nameid = False
+  for line in lines:
```

Comment:
I was puzzled at this for-loop until I saw _replace_existing_version and realized that "<namerecord nameID=...." and actual string are in separate lines.